File: src/leadforge/scrapers/nextdoor.py
import structlog

from leadforge.scrapers.base import BaseAPIClient

logger = structlog.get_logger()


class NextdoorClient(BaseAPIClient):
    """Nextdoor business scraper with cookie-based authentication."""

    def __init__(self, cookies: dict | None = None):
        super().__init__(base_url="https://nextdoor.com", timeout=30.0)
        self.cookies = cookies or {}

    async def search_business(self, name: str, zip_code: str) -> dict | None:
        """Search for a business on Nextdoor. Returns recommendation data or None."""
        if not self.cookies:
            logger.warning(
                "nextdoor_no_cookies", msg="No cookies available, skipping Nextdoor"
            )
            return None

        try:
            client = await self._get_client()
            response = await client.get(
                "/api/search/",
                params={"query": name, "type": "business"},
                cookies=self.cookies,
            )
            response.raise_for_status()
            data = response.json()

            results = data.get("results", [])
            if not results:
                return None

            biz = results[0]
            return {
                "nextdoor_id": biz.get("id"),
                "nextdoor_recommendations": biz.get("recommendation_count", 0),
                "nextdoor_neighborhood": biz.get("neighborhood"),
                "nextdoor_rating": biz.get("rating"),
            }
        except Exception as e:
            logger.warning("nextdoor_search_failed", error=str(e), name=name)
            return None

File: api/src/scrapers/nextdoor.ts
// =py scrapers/nextdoor
import { encodeParams, fetchJson } from './base';

export interface NextdoorProfile {
  nextdoor_id: string | null;
  nextdoor_recommendations: number;
  nextdoor_neighborhood: string | null;
  nextdoor_rating: number | null;
}

interface NextdoorResult {
  id?: string;
  recommendation_count?: number;
  neighborhood?: string;
  rating?: number;
}

// =py search_business — cookies are a constructor argument that no caller supplies, so this always
// returns null. Kept as an argument rather than read from COOKIE_STORE: nothing has ever exercised
// the authenticated path, and the response shape above is unverified.
export async function searchBusiness(
  name: string,
  _zipCode: string,
  cookies: Record<string, string> = {},
): Promise<NextdoorProfile | null> {
  const cookieHeader = Object.entries(cookies)
    .map(([key, value]) => `${key}=${value}`)
    .join('; ');
  if (!cookieHeader) {
    console.warn('nextdoor_no_cookies', { msg: 'No cookies available, skipping Nextdoor' });
    return null;
  }

  try {
    const params = encodeParams({ query: name, type: 'business' });
    const data = await fetchJson<{ results?: NextdoorResult[] }>(`https://nextdoor.com/api/search/?${params}`, {
      headers: { Cookie: cookieHeader },
    });

    const business = data.results?.[0];
    if (!business) return null;

    return {
      nextdoor_id: business.id ?? null,
      nextdoor_recommendations: business.recommendation_count ?? 0,
      nextdoor_neighborhood: business.neighborhood ?? null,
      nextdoor_rating: business.rating ?? null,
    };
  } catch (error) {
    console.warn('nextdoor_search_failed', { name, error: error instanceof Error ? error.message : String(error) });
    return null;
  }
}
